aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -117,6 +117,52 @@ Tensor pinverse(const Tensor& self, double rcond) {
   return V.mm(S_pseudoinv.diag().mm(U.t()));
 }
 
+static double _get_epsilon(const ScalarType& sc_type) {
+  switch (sc_type) {
+    case at::ScalarType::Half:
+      return 0.00097656;
+    case at::ScalarType::Float:
+      return 1.19209e-07;
+    case at::ScalarType::Double:
+      return 2.22044604925e-16;
+    default:
+      AT_ERROR("This function doesn't handle non-floating type");
+  }
+}
+
+Tensor _matrix_rank_helper(const Tensor& self, bool symmetric) {
+  Tensor S;
+  if (!symmetric) {
+    Tensor U, V;
+    std::tie(U, S, V) = self.svd();
+  } else {
+    Tensor eigvecs;
+    std::tie(S, eigvecs) = self.symeig();
+    S = S.abs();
+  }
+  return S;
+}
+
+int64_t matrix_rank(const Tensor& self, double tol, bool symmetric) {
+  AT_CHECK(at::isFloatingType(self.type().scalarType()) && self.dim() == 2,
+           "matrix_rank(", self.type(), "{", self.sizes(), "}): expected a 2D tensor "
+           "of floating types");
+
+  Tensor S = _matrix_rank_helper(self, symmetric);
+  return (S > tol).toType(kLong).sum().toCLong();
```

Comment:
Does it make more sense to just return the scalar tensor?

Reply:
Do you want the result to be a scalar tensor?

Reply:
It would be just a matter of removing the `toCLong()`. I thought we should stick to NumPy semantics and return an `int`.

Reply:
Yeah, but it then is a synchronization point for CUDA kernels.

Reply:
Oh I see. Sure, I will make that change.

Reply:
Thanks :)